### ISSUE TITLE
problem with varargs and following arguments

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#17188]
Consider this program:

``` chapel
proc f( x:int ...?k, named:int) {
  writeln("f with named");
}
proc f( x:int ...?k) {
  writeln("f");
}

proc main() {
  var x = 1;
  var y = 2;
  var z = 3;
  f(x,y,z);        // should print "f"
  f(x,y, named=z); // should print "f with named"
}
```

I am expecting that this program compile and run. When there is a varargs formal of variable length, I am expecting it to consume all remaining arguments (regardless of type) unless they are named arguments.

In contrast, when the varargs has fixed type, I would expect an ambiguity error, as in:

``` chapel
proc g( x:int ...2, named:int) {
  writeln("g with named");
}
proc g( x:int ...3) {
  writeln("g");
}

proc main() {
  var x = 1;
  var y = 2;
  var z = 3;
  g(x,y,z);        // ambiguity error
  g(x,y, named=z);
}
```

**Comment by bradcray** [comment]
I know we've had cases similar to this that have worked before (and, I expect, still work now).  I.e., is this specific to having the two overloads, or having the following argument share the type of the actuals, or ...?

**Comment by mppf** [comment]
The compiler seems to get tripped up only when the types match in both cases. I'm running into some test failing because with the out intent changes the `named` argument is more willing to accept more actual arguments.

**Comment by mppf** [comment]
yes - it is specific to having 2 overloads and having the following argument share the type of the actuals (or be generic).

**Comment by mppf** [comment]
I'm running into this in PR #16990 with test/types/records/ferguson/tuples/const1.chpl. I think I will future-ize that test for now to allow fixing this to be a separate effort, since it's arguably an existing issue.

**Comment by bradcray** [comment]
Thanks for the additional info, that works for me.

**Comment by bradcray** [comment]
I wonder if we could fix this in dyno.  Or maybe we should even try to fix it sooner.  @benharsh and I ran back into it in #18504

**Comment by benharsh** [comment]
Issue #23528 would benefit from some progress in this area.

@mppf: Do you still think the first call to ``f`` should be unambiguous? Personally, it's not clear to me that that should be the case. 

One possible way to approach this issue could be the introduction of a language feature that indicates an argument must be named in the call, and is otherwise not considered when binding positional arguments.

Python3 seems to be pursuing this direction based on [PEP 3102](https://peps.python.org/pep-3102/) with their "keyword arguments". I don't think the syntax there is right for Chapel, but the idea seems like it could help reduce or eliminate ambiguity in this situation.

In the writeln/serializers use-case I tend to think about the feature in terms of "configuring" the method's behavior, so perhaps the ``config`` keyword could play a role here?

```chpl
proc foo( x:int ...?k, config other:int) {
}

foo(1, 2, 3); // error: 'config' argument 'other' was not named and does not have a default value
foo(1, 2, other=3); // OK

proc bar( args: int...?k, config optional : int = -1) {
  write("args: "); for a in args do write(a, " "); writeln();
  writeln("optional: ", optional);
}

// args: 1 2 3 
// optional: -1
bar(1, 2, 3);
```

In this idea ``config`` could be a kind of modifier on top of the other argument intents, so you could write ``config in`` or ``config ref`` as well.

**Comment by bradcray** [comment]
> Do you still think the first call to f should be unambiguous?

Though I'm not @mppf, I'd be very comfortable with that.  Specifically, my intuition is that a varargs should be greedy in matching arguments until it gets to one it can't handle.

Taking your similar-but-different case in #23528, I react similarly:

```chapel
proc foo(args...?k, optional:? = new X()) {
```

First, this case seems simpler because it's just one routine rather than two overloads, so no need to worry about ambiguity.  And then, since `args` is a "gobble anything" varargs and `optional` is defaulted, I'd say that the right behavior is for `args` to match everything unless an `optional=` argument is provided.

**Comment by mppf** [comment]
> > Do you still think the first call to f should be unambiguous?
> 
> Though I'm not @mppf, I'd be very comfortable with that.

IMO it's reasonable to define the language this way. I think my justification for it would be that optional arguments should be optional, and the way it works today, we can't have varargs + optional arguments that are optional. We would be giving up varargs + non-optional trailing arguments, but I'm not convinced that case is used or useful. Or, we could make the gobbling stop only for optional arguments.

Would it be a breaking change? I think we could claim it is a bug, at least if it does not trip up user codes. The spec section https://chapel-lang.org/docs/language/spec/procedures.html#variable-number-of-arguments doesn't seem to say anything about whether or not it's even possible to have more formal arguments match after an unbounded varargs.

@bradcray 

> Specifically, my intuition is that a varargs should be greedy in matching arguments until it gets to one it can't handle.

What can stop a greedy varargs from matching an argument? I can think of two things:
 1. It runs out of size, as with e.g. `proc g( x:int ...3)` when called like `f(1,2,3,4)`. Of course, maybe we shouldn't call this pattern varargs.
 2. It runs into a named argument.

Specifically, I don't think we want to have the varargs gobbling process depend on the types. E.g. `proc h(args:int ..., other: string)` `h(1,2,3, "hello")`. I think such a thing is technically possible but a departure from what we do today that I don't think is well-motivated at this point.

**Comment by bradcray** [comment]
> I don't think we want to have the varargs gobbling process depend on the types.

This is what I was imagining, yeah.  It'd be fine with me if we didn't support it for now since it'd be a non-breaking change to add support for it later (and I think the code is clearer when using named arguments in such cases anyway).

> It runs into a named argument.

Do we not still do named argument matches first, and others afterwards?  That's what I was imagining in a case like the OP—that `named` would be found and matched first, that remaining arguments would be matched positionally, and they'd find that the positional argument they matched happened to be a compatible varargs.

**Comment by mppf** [comment]
> This is what I was imagining, yeah (that once a typed vararg formal stopped seeing arguments that matched its type, it would stop, and the argument matching algorithm would proceed to the next formal, if any). That said, it'd also be fine with me if we didn't support this behavior for now since (I believe) it'd be a non-breaking change to add support for it later (and I think the code is clearer when using named arguments in such cases anyway).

I'm not so sure it'd be a non-breaking change to go from varargs gobbling not considering types to varargs gobbling considering types. Consider this example:

``` chapel
proc f(args: int ..., arg: string="one") { }  // f#1
proc f(a: int, b: int, arg: string="two") { } // f#2
f(1,2,"hello");
```

If the varargs doesn't stop based on the type of the arguments, then I think the compiler would consider f#1 as a candidate, instantiate it for varargs with size 3, but then reject the candidate because it can't pass `"hello"` to an int. So the call would resolve to f#2. But, if the compiler were then to consider types when deciding when to stop allocating actuals to the vararg, it would become ambiguous.

**Comment by benharsh** [comment]
I think some of my apprehension around this proposed change is related to having recently re-implemented varargs in dyno, and basing that implementation on the production compiler's behavior. My impression at the time was that the production compiler intentionally tried to be non-greedy with varargs, and that the current behavior wasn't just some trivial consequence of a particular implementation, and so "we" wanted it at some point in the past. But there are a number of assumptions and behaviors embedded in the production compiler at some point that haven't necessarily received due attention.

I think there are a couple of useful steps to take going forward:

1) Look at how other languages handle this situation, and whether their varargs are greedy
2) Try prototyping this change and see what tests/modules are impacted
    i. See if the prototype allows us to implement the desired optional-serializer argument for write/writeln

**Comment by bradcray** [comment]
> the current behavior wasn't just some trivial consequence of a particular implementation, and so "we" wanted it at some point in the past.

I definitely remember talking in the early days about how support for named args was a nice tool for supporting optional arguments that follow varargs, so it would surprise me if that turned out to be the case.  But my memory's far from perfect...  I like your proposal (where I think we should focus on languages that permit `a=1`-style actual arguments... e.g., I don't know that we can learn much from C).

**Comment by jabraham17** [comment]
I was talking with Brad today and we realized I had a workaround for this issue (being used in the Python module). The below code example works as the OP expects it to

```chapel
pragma "last resort"
proc f( x:int ...?k, named:int) {
  writeln("f with named");
}
proc f( x:int ...?k) {
  writeln("f");
}

proc main() {
  var x = 1;
  var y = 2;
  var z = 3;
  f(x,y,z);        // should print "f"
  f(x,y, named=z); // should print "f with named"
}
```

With a little more code, you can make just the `f(named=z)` case work as well

```chapel
pragma "last resort"
proc f( x:int ...?k, named:int) {
  writeln("f with named");
}
pragma "last resort"
proc f(named:int) {
  writeln("f with named2");
}
proc f( x:int ...?k) {
  writeln("f");
}

proc main() {
  var x = 1;
  var y = 2;
  var z = 3;
  f(x,y,z);        // should print "f"
  f(x,y, named=z); // should print "f with named"
  f(z);  // should print "f"
  f(named=z); // should print "f with named2"
}

```

This is not to say the bug no longer exists, rather that we have workarounds available to make it work today
